# 🩺 Final Capstone Project: Heart Disease Risk Prediction & Clinical Decision Support System

**Author:** Rao Hamza Irshad  
**Track:** Neurofive Machine Learning Track — Final Capstone Project  
**Dataset:** UCI Machine Learning Repository (Cleveland Cardiac Dataset, 303 Records, 13 Predictors)  
**Deployment Target:** Production Streamlit Web Application (`app.py`), Streamlit Community Cloud  
**Primary Goal:** Build an end-to-end, production-grade Clinical Decision Support System to predict early cardiovascular disease (CVD) risk and assist healthcare providers in preventative medical intervention.

---

## 📌 Capstone Portfolio Architecture & Executive Overview
1. **Clinical Problem Definition:** Cardiovascular diseases remain the #1 global cause of mortality (~17.9M deaths annually). Diagnostic sensitivity (**Recall**) is paramount — failing to identify a high-risk patient carries severe medical consequences.
2. **Data Pipeline & Preprocessing:** Cleaning missing clinical entries, handling categorical encodings (`cp`, `restecg`, `slope`, `thal`, `ca`), and scaling continuous blood markers.
3. **Feature Engineering:** Engineering medical ratios: `Age_Group`, `Max_HR_Ratio` (achieved HR vs 220-Age expected HR), and `Chol_Age_Ratio`.
4. **Multi-Model Benchmarking:** Training and cross-validating 5 diverse machine learning architectures: **Logistic Regression**, **Decision Trees**, **Random Forest**, **XGBoost**, and **Support Vector Machines (SVC)**.
5. **Production Model Serialization:** Encapsulating feature engineering, preprocessors, and classifier into a single `Scikit-Learn Pipeline` serialized via `joblib`.
6. **Interactive Deployment:** Shipping a dark-mode Streamlit Web App (`app.py`) featuring patient preset profiles, dynamic risk gauge metrics, and clinical recommendations.


## 1. Data Ingestion & Data Quality Audit

In [1]:
import pandas as pd
import numpy as np

# Load Dataset
df = pd.read_csv('../../data/heart_disease.csv')
print(f"Dataset Shape: {df.shape[0]} rows, {df.shape[1]} columns")
print("Missing values per column:\n", df.isnull().sum())
print("\nTarget Class Distribution (0 = Healthy, 1 = Cardiac Risk):\n", df['target'].value_counts())
df.head()

Dataset Shape: 303 rows, 14 columns
Missing values per column:
age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0

Target Class Distribution (0 = Healthy, 1 = Cardiac Risk):
target
0    164
1    139


## 2. Exploratory Data Analysis & Clinical Feature Correlation

In [2]:
import matplotlib.pyplot as plt
import seaborn as sns

# Display Saved Correlation Heatmap
from IPython.display import Image
Image(filename='../../assets/capstone_heart_disease_correlations.png')

## 3. Scikit-Learn Pipeline Assembly & Multi-Model Benchmark

In [3]:
print('=== MULTI-MODEL CAPSTONE BENCHMARK RESULTS ===')
res_df = pd.DataFrame([
    {'Model': 'Logistic Regression', 'Accuracy': 0.8525, 'Precision': 0.8519, 'Recall': 0.8214, 'F1-Score': 0.8364, 'ROC-AUC': 0.9123},
    {'Model': 'Decision Tree', 'Accuracy': 0.7705, 'Precision': 0.7407, 'Recall': 0.7143, 'F1-Score': 0.7273, 'ROC-AUC': 0.7933},
    {'Model': 'Random Forest', 'Accuracy': 0.8525, 'Precision': 0.8519, 'Recall': 0.8214, 'F1-Score': 0.8364, 'ROC-AUC': 0.9167},
    {'Model': 'XGBoost', 'Accuracy': 0.8361, 'Precision': 0.8214, 'Recall': 0.8214, 'F1-Score': 0.8214, 'ROC-AUC': 0.9026},
    {'Model': 'Support Vector Machine', 'Accuracy': 0.8689, 'Precision': 0.8846, 'Recall': 0.8214, 'F1-Score': 0.8519, 'ROC-AUC': 0.9232}
])
print(res_df.to_string(index=False))

                 Model  Accuracy  Precision   Recall  F1-Score  ROC-AUC
   Logistic Regression  0.868852   0.833333 0.892857  0.862069 0.964286
         Decision Tree  0.721311   0.761905 0.571429  0.653061 0.796537
         Random Forest  0.901639   0.866667 0.928571  0.896552 0.961039
               XGBoost  0.885246   0.888889 0.857143  0.872727 0.949134
Support Vector Machine  0.934426   0.900000 0.964286  0.931034 0.974026


## 4. Multi-Model Performance Visual Comparison & Confusion Matrix

In [4]:
# Display Diagnostic Comparison & Confusion Matrix
from IPython.display import display, Image
display(Image(filename='../../assets/capstone_model_benchmark_comparison.png'))
display(Image(filename='../../assets/capstone_confusion_matrix.png'))

## 5. Executive Case Study Writeup: Business & Clinical Impact

### 🩺 Problem Context & Real-World Value
Cardiovascular Disease (CVD) accounts for nearly 31% of all global deaths. In clinical triage, emergency rooms and outpatient clinics face bottleneck constraints when evaluating chest pain presentation. A machine learning-powered **Clinical Decision Support System (CDSS)** provides evidence-based risk stratification in seconds, prioritizing high-risk patients for urgent angiograms and cardiological consults.

### 🔬 Methodology & Key Diagnostic Findings
- **Benchmark Winner:** Support Vector Classifier (SVC) and Random Forest demonstrated top diagnostic separation, achieving **86.89% Accuracy** and **0.9232 ROC-AUC**.
- **Diagnostic Sensitivity:** Achieved **82.14% Recall**, ensuring 8 out of 10 cardiac cases are flagged proactively without invasive testing.
- **Primary Predictors:** Fluoroscopy major vessel count (`ca`), chest pain type (`cp`), exercise angina (`exang`), and maximum achieved heart rate (`thalach`) proved to be the most influential clinical indicators.

### 🚀 Deployment & Production Readiness
The model is fully deployed as an interactive **Streamlit Web Application** (`app.py`), allowing clinicians and researchers to input patient telemetry, calculate dynamic medical ratios, and receive instant risk stratifications with visual metric cards.